server/src/main/java/org/elasticsearch/index/translog/AsyncTranslog.java

```diff
@@ -0,0 +1,249 @@
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.AbstractIndexShardComponent;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.AsyncTranslogEventDispatcher.OperationEvent;
+import org.elasticsearch.index.translog.Translog.Location;
+import org.elasticsearch.index.translog.Translog.Operation;
+import org.elasticsearch.index.translog.Translog.Snapshot;
+import org.elasticsearch.index.translog.Translog.TranslogGeneration;
+
+/**
+ * use async thanslog by disruptor to decompose method add (synchronized method) and other method(eg. sync)
+ *
+ * @author wesleydeng@tencent.com
+ */
+public class AsyncTranslog extends AbstractIndexShardComponent implements ITranslog {
+
+    private ITranslog translog;
+
+    private AsyncTranslogEventDispatcher dispatcher;
+
+    private AtomicLong unConsumeCounter;
+
+
+    public AsyncTranslog(ITranslog translog, ShardId shardId,
+        IndexSettings indexSettings, AsyncTranslogEventDispatcher dispatcher) {
+        super(shardId, indexSettings);
+        this.translog = translog;
+        this.dispatcher = dispatcher;
+        this.unConsumeCounter = new AtomicLong(0);
+
+    }
+
+    @Override
+    public boolean isOpen() {
+        return translog.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        translog.close();
+    }
+
+    @Override
+    public Path location() {
+        return translog.location();
+    }
+
+    @Override
+    public long currentFileGeneration() {
+        return translog.currentFileGeneration();
+    }
+
+    @Override
+    public long getMinFileGeneration() {
+        return translog.getMinFileGeneration();
+    }
+
+    @Override
+    public int totalOperations() {
+        return translog.totalOperations();
+    }
+
+    @Override
+    public long sizeInBytes() {
+        return translog.sizeInBytes();
+    }
+
+    @Override
+    public int totalOperationsByMinGen(long minGeneration) {
+        return translog.totalOperationsByMinGen(minGeneration);
+    }
+
+    @Override
+    public int estimateTotalOperationsFromMinSeq(long minSeqNo) {
+        return translog.estimateTotalOperationsFromMinSeq(minSeqNo);
+    }
+
+    @Override
+    public long sizeInBytesByMinGen(long minGeneration) {
+        return translog.sizeInBytesByMinGen(minGeneration);
+    }
+
+    /**
+     * location is used for sync on request or search in translog. but AsyncTranslog is userd for aysnc
+     *
+     * @return always return null
+     * @throws IOException @see org.elasticsearch.index.translog.Translog#add
+     */
+    @Override
+    public Location add(Operation operation) throws IOException {
+        //produce using ring buffer
+        dispatcher.produce(new OperationEvent(operation, this.translog, this.unConsumeCounter));
+        //async return no location
+        return null;
```

Comment:
This breaks [wait_for_refresh](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html#refresh_wait_for-force-refresh) and other refresh features.

Reply:
> This breaks [wait_for_refresh](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html#refresh_wait_for-force-refresh) and other refresh features.

As I know, wait_for_refresh is not suitable for the scene that require high indexing performance, because it will result in writing thread waiting and smaller segement (which will in trun increase time to query doc version when indexing).

So. Is it really necessary to force return the low level info( offset of Translog file)  in async durability?